File: clients/agent.js
// Get the agentId and play mode from the command-line arguments
// node agent.js ID playAsPlayer
// process.argv[0] is node
// process.argv[1] is client1.js
// process.argv[2] is ID
// process.argv[3] is play mode

const WebSocket = require("ws");

// Configuration
const CONFIG = {
  defaultId: "k",
  defaultMode: "up",
  validModes: [
    "up",
    "down",
    "left",
    "right",
    "random",
    "timeout",
    "apple",
    "survive",
  ],
  validDirections: ["up", "down", "left", "right"],
  baseDelay: 100,
  wsUrl: "ws://localhost:3000",
};

// Game state
const gameState = {
  agentId: process.argv[2] || CONFIG.defaultId,
  agentMode: process.argv[3],
  delayBetweenMoves: CONFIG.baseDelay,
  playerName: null, // Add this to store the player name
};

// Initialize agent mode
if (!gameState.agentMode || !CONFIG.validModes.includes(gameState.agentMode)) {
  console.error(
    "Mode not provided or invalid, using default:",
    CONFIG.defaultMode
  );
  gameState.agentMode = CONFIG.defaultMode;
}

// Movement helpers
const movementHelpers = {
  getNextPosition(current, direction) {
    const pos = { x: current.x, y: current.y };
    const moves = {
      up: () => (pos.x -= 1),
      down: () => (pos.x += 1),
      left: () => (pos.y -= 1),
      right: () => (pos.y += 1),
    };
    moves[direction]();
    return pos;
  },

  findPlayerHead(map, playerName) {
    if (!map || !playerName) return { x: 0, y: 0 };

    for (let i = 0; i < map.length; i++) {
      for (let j = 0; j < map[i].length; j++) {
        const cell = map[i][j];
        if (
          cell &&
          cell.type === "snake-head" &&
          cell.playerName === playerName
        ) {
          return { x: i, y: j };
        }
      }
    }

    console.error("Player not found in map");
    // Return default position if player not found
    return { x: 0, y: 0 };
  },

  isSafeMove(map, pos) {
    if (
      !map ||
      !map[0] ||
      pos.x < 0 ||
      pos.x >= map.length ||
      pos.y < 0 ||
      pos.y >= map[0].length
    ) {
      return false;
    }

    const cell = map[pos.x][pos.y];

    // If cell is null, it's safe
    if (cell === null) return true;

    // If it's a snake part or border, it's not safe
    if (
      cell.type === "snake-head" ||
      cell.type === "snake-body" ||
      cell.type === "border"
    ) {
      return false;
    }

    // check if any cell next to pos is border or enemy player head
    ["border", "snake-head"].forEach((cellType) => {
      if (
        (map[pos.x - 1] &&
          map[pos.x - 1][pos.y] &&
          map[pos.x - 1][pos.y].type === cellType) ||
        (map[pos.x + 1] &&
          map[pos.x + 1][pos.y] &&
          map[pos.x + 1][pos.y].type === cellType) ||
        (map[pos.x][pos.y - 1] && map[pos.x][pos.y - 1].type === cellType) ||
        (map[pos.x][pos.y + 1] && map[pos.x][pos.y + 1].type === cellType)
      ) {
        return false;
      }
    });

    // Check if it's a safe item type
    const safeTypes = [
      "apple",
      "armour",
      "freeze",
      "golden-apple",
      "katana",
      "nausea",
      "reset-borders",
      "shorten",
      "tron",
    ];

    return safeTypes.includes(cell.type);
  },
};

// Strategy implementations
const strategies = {
  findSafeDirection(map, playerHead) {
    if (!map || !playerHead) {
      return CONFIG.validDirections[
        Math.floor(Math.random() * CONFIG.validDirections.length)
      ];
    }

    // Define all possible moves with their priorities
    const directions = [
      { dRow: -1, dColumn: 0, move: "up" },
      { dRow: 1, dColumn: 0, move: "down" },
      { dRow: 0, dColumn: -1, move: "left" },
      { dRow: 0, dColumn: 1, move: "right" },
    ];

    // Randomize direction order to avoid predictable patterns
    directions.sort(() => Math.random() - 0.5);

    // First pass: Check for immediate safe moves
    for (const dir of directions) {
      const nextPos = {
        x: playerHead.x + dir.dRow,
        y: playerHead.y + dir.dColumn,
      };

      if (movementHelpers.isSafeMove(map, nextPos)) {
        return dir.move;
      }
    }

    // If no safe moves found, pick a random direction
    return CONFIG.validDirections[
      Math.floor(Math.random() * CONFIG.validDirections.length)
    ];
  },

  findClosestApple(map, playerHead) {
    const rows = map.length;
    const cols = map[0].length;
    const queue = [[playerHead.x, playerHead.y, []]];
    const visited = new Set();

    while (queue.length > 0) {
      const [x, y, path] = queue.shift();
      const key = `${x},${y}`;

      if (visited.has(key)) continue;
      visited.add(key);

      const cell = map[x][y];
      if (cell && cell.type === "apple") {
        // Validate the entire path to make sure it's still safe
        let currentPos = { x: playerHead.x, y: playerHead.y };
        for (const move of path) {
          const nextPos = movementHelpers.getNextPosition(currentPos, move);
          if (!movementHelpers.isSafeMove(map, nextPos)) {
            return null; // Path is not safe anymore
          }
          currentPos = nextPos;
        }
        return path;
      }

      const directions = [
        { dRow: -1, dColumn: 0, move: "up" },
        { dRow: 1, dColumn: 0, move: "down" },
        { dRow: 0, dColumn: -1, move: "left" },
        { dRow: 0, dColumn: 1, move: "right" },
      ];

      for (const { dRow, dColumn, move } of directions) {
        const newX = x + dRow;
        const newY = y + dColumn;
        const nextPos = { x: newX, y: newY };

        if (newX < 0 || newX >= rows || newY < 0 || newY >= cols) continue;

        // Use isSafeMove to check if the next position is safe
        if (!movementHelpers.isSafeMove(map, nextPos)) continue;

        queue.push([newX, newY, [...path, move]]);
      }
    }
    return null;
  },
};

// Movement decision logic
function decideNextMove(map, mode) {
  const playerHead = movementHelpers.findPlayerHead(map, gameState.playerName);

  switch (mode) {
    case "timeout":
      return strategies.findSafeDirection(map, playerHead);

    case "survive":
      return strategies.findSafeDirection(map, playerHead);

    case "apple": {
      const path = strategies.findClosestApple(map, playerHead);
      if (path && path.length > 0) {
        const nextMove = path[0];
        const nextPos = movementHelpers.getNextPosition(playerHead, nextMove);
        return movementHelpers.isSafeMove(map, nextPos)
          ? nextMove
          : strategies.findSafeDirection(map, playerHead);
      }
      return strategies.findSafeDirection(map, playerHead);
    }

    case "random":
      return CONFIG.validDirections[
        Math.floor(Math.random() * CONFIG.validDirections.length)
      ];

    default:
      return mode;
  }
}

// WebSocket setup and event handlers
const ws = new WebSocket(`${CONFIG.wsUrl}?id=${gameState.agentId}`);

ws.on("open", () => console.log("Connected to WebSocket server"));
ws.on("error", (error) => console.error("WebSocket error:", error));
ws.on("close", () => console.log("Disconnected from WebSocket by server"));

// Update WebSocket message handler
ws.on("message", (data) => {
  const receivedMsg = JSON.parse(data.toString("utf-8"));

  // Store player name when receiving the connection success message
  if (receivedMsg.message === "Player connected successfully.") {
    console.log(
      "Agent connected with name: '" +
        receivedMsg.name +
        "' and id: '" +
        gameState.agentId +
        "'."
    );
    gameState.playerName = receivedMsg.name;
    return;
  }

  const gameIsOver = receivedMsg.winner !== null;

  if (!gameIsOver && receivedMsg.map) {
    const direction = decideNextMove(receivedMsg.map, gameState.agentMode);

    const move = { playerId: gameState.agentId, direction };

    setTimeout(() => {
      ws.send(JSON.stringify(move));
      if (gameState.agentMode === "timeout") {
        gameState.delayBetweenMoves += 100;
      }
    }, gameState.delayBetweenMoves);
  }
});

File: clients/agent.py
import asyncio
import websockets
import json
import sys
import random
import time

# Default values
DEFAULT_AGENT_ID = "l"
VALID_DIRECTIONS = ["up", "down", "left", "right"]
VALID_MODES = ["up", "down", "left", "right", "random", "timeout", "apple", "survive"]
BASE_DELAY = 0.05  # 50ms

# Add helper functions for apple and survive modes
def find_player_head(game_map, player_symbol):
    for i in range(len(game_map)):
        for j in range(len(game_map[i])):
            if game_map[i][j] and game_map[i][j].get('type') == 'snake-head' and game_map[i][j].get('player') == player_symbol.lower():
                return {"x": i, "y": j}
    return {"x": 0, "y": 0}

def is_safe_move(game_map, pos):
    if (pos["x"] < 0 or pos["x"] >= len(game_map) or 
        pos["y"] < 0 or pos["y"] >= len(game_map[0])):
        return False
    cell = game_map[pos["x"]][pos["y"]]
    return cell is None or (cell and cell.get('type') == 'apple')

def find_safe_direction(game_map, player_head):
    directions = [
        {"dx": -1, "dy": 0, "move": "up"},
        {"dx": 1, "dy": 0, "move": "down"},
        {"dx": 0, "dy": -1, "move": "left"},
        {"dx": 0, "dy": 1, "move": "right"}
    ]
    random.shuffle(directions)

    for direction in directions:
        new_pos = {
            "x": player_head["x"] + direction["dx"],
            "y": player_head["y"] + direction["dy"]
        }
        if is_safe_move(game_map, new_pos):
            return direction["move"]
    
    return random.choice(VALID_DIRECTIONS)

def find_closest_apple(game_map, player_head):
    rows, cols = len(game_map), len(game_map[0])
    queue = [(player_head["x"], player_head["y"], [])]
    visited = set()

    while queue:
        x, y, path = queue.pop(0)
        key = f"{x},{y}"

        if key in visited:
            continue
        visited.add(key)

        cell = game_map[x][y]
        if cell and cell.get('type') == 'apple':
            return path

        directions = [
            {"dx": -1, "dy": 0, "move": "up"},
            {"dx": 1, "dy": 0, "move": "down"},
            {"dx": 0, "dy": -1, "move": "left"},
            {"dx": 0, "dy": 1, "move": "right"}
        ]

        for direction in directions:
            new_x = x + direction["dx"]
            new_y = y + direction["dy"]

            if new_x < 0 or new_x >= rows or new_y < 0 or new_y >= cols:
                continue

            cell = game_map[new_x][new_y]
            if (cell is not None and 
                cell.get('type') != 'apple' and 
                (cell.get('type') == 'snake-head' or cell.get('type') == 'snake-body')):
                continue

            queue.append((new_x, new_y, path + [direction["move"]]))
    
    return None

async def connect_to_game_server(agent_id, mode):
    uri = f"ws://localhost:3000?id={agent_id}"
    delay = BASE_DELAY

    async with websockets.connect(uri) as websocket:
        print("Connected to WebSocket server")
        
        # Receive initial response
        response = await websocket.recv()
        print(f"Initial server response: {response}")

        try:
            while True:
                # Receive game state
                message = await websocket.recv()
                game_state = json.loads(message)
                # print(f"Received game state: {game_state}")

                # Check if game is over
                if game_state.get('winner') is not None:
                    print("Game over!")
                    break

                # Enhanced move determination
                direction = mode
                if game_state.get('map'):
                    player_head = find_player_head(game_state['map'], agent_id.upper())
                    
                    if mode == "survive":
                        direction = find_safe_direction(game_state['map'], player_head)
                    elif mode == "apple":
                        path = find_closest_apple(game_state['map'], player_head)
                        if path and path[0]:
                            next_pos = {
                                "x": player_head["x"],
                                "y": player_head["y"]
                            }
                            if path[0] == "up": next_pos["x"] -= 1
                            elif path[0] == "down": next_pos["x"] += 1
                            elif path[0] == "left": next_pos["y"] -= 1
                            elif path[0] == "right": next_pos["y"] += 1
                            
                            direction = path[0] if is_safe_move(game_state['map'], next_pos) else find_safe_direction(game_state['map'], player_head)
                        else:
                            direction = find_safe_direction(game_state['map'], player_head)
                    elif mode == "random":
                        direction = random.choice(VALID_DIRECTIONS)

                move = {
                    "playerId": agent_id,
                    "direction": direction
                }

                # Add delay and increase if in timeout mode
                await asyncio.sleep(delay)
                if mode == "timeout":
                    delay += 0.05  # Add 50ms each move

                # Send move
                await websocket.send(json.dumps(move))
                print(f"Sent move: {move}")

        except websockets.exceptions.ConnectionClosed:
            print("Disconnected from WebSocket server")
        except Exception as e:
            print(f"Error: {e}")

def main():
    # Get command line arguments
    agent_id = sys.argv[1] if len(sys.argv) > 1 else DEFAULT_AGENT_ID
    mode = sys.argv[2] if len(sys.argv) > 2 else "up"

    # Validate mode
    if mode not in VALID_MODES:
        print(f"Invalid mode: {mode}. Using default: up")
        mode = "up"

    print(f"Starting agent with ID: {agent_id}, Mode: {mode}")
    
    # Run the client
    asyncio.run(connect_to_game_server(agent_id, mode))

if __name__ == "__main__":
    main()